drivers/i2c/i2c_esp32.c

```diff
@@ -0,0 +1,781 @@
+	u8_t i;
+	int r = 0;
+
+	k_sem_take(&data->transfer_sem, K_FOREVER);
+
+	/* Mask out unused address bits, and make room for R/W bit */
+	addr = (addr & BIT_MASK(data->config.bits.use_10_bit_addr ? 10 : 7)) << 1;
+
+	for (i = 0; i < num_msgs; i++) {
+		if ((msgs[i].flags & I2C_MSG_RW_MASK) == I2C_MSG_WRITE) {
+			r = i2c_esp32_write_msg(dev, addr, msgs[i]);
+		} else {
+			r = i2c_esp32_read_msg(dev, addr, msgs[i]);
+		}
+
+		if (r < 0) {
+			break;
+		}
+	}
+
+	k_sem_give(&data->transfer_sem);
+
+	return r;
+}
+
+static void i2c_esp32_isr(void *arg)
+{
+	const int fifo_give_mask = I2C_ACK_ERR_INT_ST |
+				   I2C_TIME_OUT_INT_ST |
+				   I2C_TRANS_COMPLETE_INT_ST |
+				   I2C_ARBITRATION_LOST_INT_ST;
+	struct device *device = arg;
+	const struct i2c_esp32_config *config = device->config->config_info;
+
+	if (*config->regs.intr_status & fifo_give_mask) {
+		struct i2c_esp32_data *data = device->driver_data;
+
+		/* Only give the semaphore if a watched interrupt happens.
+		 * Error checking is performed at the other side of the
+		 * semaphore, by reading the status register.
+		 */
+		k_sem_give(&data->fifo_sem);
+	}
+
+	/* Acknowledge all interrupts, even non-watched ones */
+	i2c_esp32_ack_intrs(config);
+}
+
+static int i2c_esp32_init(struct device *dev);
+
+static const struct i2c_driver_api i2c_esp32_driver_api = {
+	.configure = i2c_esp32_configure,
+	.transfer = i2c_esp32_transfer,
+};
+
+#ifdef CONFIG_I2C_0
+static const struct i2c_esp32_config i2c_esp32_config_0 = {
+	.sig = {
+		.sda_out = I2CEXT0_SDA_OUT_IDX,
+		.sda_in = I2CEXT0_SDA_IN_IDX,
+		.scl_out = I2CEXT0_SCL_OUT_IDX,
+		.scl_in = I2CEXT0_SCL_IN_IDX,
+	},
+	.pins = {
+		.scl = CONFIG_I2C_ESP32_0_SCL_PIN,
+		.sda = CONFIG_I2C_ESP32_0_SDA_PIN,
+		.sda_pull_up =
+			IS_ENABLED(CONFIG_I2C_ESP32_0_SDA_PULL_UP),
+		.scl_pull_up =
+			IS_ENABLED(CONFIG_I2C_ESP32_0_SCL_PULL_UP),
+		.default_level =
+			IS_ENABLED(CONFIG_I2C_ESP32_0_DEFAULT_LEVEL),
+	},
+	.enable_bits = {
+		.clk = DPORT_I2C_EXT0_CLK_EN,
+		.rst = DPORT_I2C_EXT0_RST,
+	},
+	.regs = {
+		.ctr = (u32_t *)I2C_CTR_REG(0),
+		.fifo_cfg = (u32_t *)I2C_FIFO_CONF_REG(0),
+		.fifo = (u32_t *)I2C_DATA_APB_REG(0),
+		.slave_addr = (u32_t *)I2C_SLAVE_ADDR_REG(0),
+		.intr = (u32_t *)I2C_INT_ENA_REG(0),
+		.intr_clr = (u32_t *)I2C_INT_CLR_REG(0),
+		.intr_status = (u32_t *)I2C_INT_STATUS_REG(0),
+		.intr_raw = (u32_t *)I2C_INT_RAW_REG(0),
+		.status = (u32_t *)I2C_SR_REG(0),
+		.cmd_base = (u32_t *)I2C_COMD0_REG(0),
+		.timeout = (u32_t *)I2C_TO_REG(0),
+		.scl_low_period = (u32_t *)I2C_SCL_LOW_PERIOD_REG(0),
+		.scl_high_period = (u32_t *)I2C_SCL_HIGH_PERIOD_REG(0),
+		.scl_start_h_time = (u32_t *)I2C_SCL_START_HOLD_REG(0),
+		.scl_rstart_s_time = (u32_t *)I2C_SCL_RSTART_SETUP_REG(0),
+		.scl_stop_h_time = (u32_t *)I2C_SCL_STOP_HOLD_REG(0),
+		.scl_stop_s_time = (u32_t *)I2C_SCL_STOP_SETUP_REG(0),
+		.sda_hold_time = (u32_t *)I2C_SDA_HOLD_REG(0),
+		.sda_sample_time = (u32_t *)I2C_SDA_SAMPLE_REG(0),
+	},
```

Comment:
Yeah, this is the 76-byte mess I was complaining about when I saw the struct definition.  I just don't see the point, why not make sure there's a bus number available everywhere in code where you have a handle to the I2C device and just use those macros directly?  I don't see that this adds anything but extra code and rodata, for no benefit.

Reply:
I did the same thing in the GPIO driver, but things got out of hand for I2C, yeah.  Will store just `0` or `1` and use these registers directly.

Reply:
CMSIS standard used by the ARM processors solves this issue nicely using define macros. You can have a look at the ext/hal/atmel/asf/sam/include/sam3x/component/twi.h file and then ext/hal/atmel/asf/sam/include/sam3x/sam3x8e.h where TWI0 and TWI1 (I2C modules on Atmel SoCs) are defined as 
```
#define TWI0       ((Twi    *)0x4008C000U) /**< \brief (TWI0      ) Base Address */
#define TWI1       ((Twi    *)0x40090000U) /**< \brief (TWI1      ) Base Address */
```
You can check how this is used in the code by looking at the Atmel SAM I2C driver at drivers/i2c/twihs_sam.c. Maybe you find it helpful.